### PR TITLE
re-add unet refactor PR 

### PR DESCRIPTION
**Pull request by yiyixuxu**
we accidentally reverted this PR in a merge conflict, adding it back 
https://github.com/huggingface/diffusers/pull/6478

### CODE REVIEW

**Comment by yiyixuxu** [comment]
@ultranity 
I'm really sorry that we accidentally reverted your UNet refactor PR and only found it out now .... 
I added it back here and made you the author. Would you be kind enough to review it again? 

We will be more careful in the future and make sure this won't happen again.

**Comment by ultranity** [comment]
Looks good for me.

**Review by DN6 — APPROVED** [review]
LGTM 👍🏽

File: src/diffusers/models/unets/unet_2d_condition.py [hunk 1]
```diff
@@ -599,14 +469,214 @@ def __init__(
...
+
+        self._set_pos_net_if_use_gligen(attention_type=attention_type, cross_attention_dim=cross_attention_dim)
+
+    def _check_config(
+        self,
+        down_block_types: Tuple[str],
+        up_block_types: Tuple[str],
+        only_cross_attention: Union[bool, Tuple[bool]],
+        block_out_channels: Tuple[int],
+        layers_per_block: [int, Tuple[int]],
```
Comment (a-r-r-o-w) [inline src/diffusers/models/unets/unet_2d_condition.py:488]: @yiyixuxu Seems like an invalid type annotation... My linter breaks when viewing this file :(

Comment (yiyixuxu) [reply src/diffusers/models/unets/unet_2d_condition.py:488]: ohh yes! do you want to open a PR?

**Pull request merged**